### PR TITLE
Lower interpolation into a call to concat

### LINKED ISSUE #16247: Optimization for string interpolation - simple scenarios: unfolding constants and lowering to concatenation
Seems like there is a lot we could do to improve perf of string interpolation. One good example is https://github.com/fsharp/fslang-suggestions/issues/1108, but also some low hanging fruits to start with could be unfolding compile time constants, e.g.:
```fsharp
[<Literal>]
let x = "hello"

[<Literal>]
let y = "world"

let z = $"{x} {y}!"
```
in the above, the last line could be lowered to `let z = "hello world!"`

Another one could be lowering to `System.String.Concat` when it applies, e.g. `let f (x: string) (y: string) = $"{x},{y}."` could be rewritten as `let f (x: string) (y: string) = System.String.Concat(x, ",", y, ".")`, because we know that all expression holes are filled with objects of type string and there are at most 4 pieces to concatenate.

### PR DESCRIPTION
**Pull request by abonie**
## Description

Optimization that lowers string interpolation into a call to concat iff there are at most 4 string parts and all fill expressions are strings.

Fixes #16247

## Benchmarks

Run a benchmark like in [this gist](https://gist.github.com/abonie/af32962df723a4a53a4e96ed84eb1f88) with and without the feature flag set.

Results without the flag (**no optimization**):
```

BenchmarkDotNet v0.13.12, Windows 11 (10.0.22621.3155/22H2/2022Update/SunValley2)
11th Gen Intel Core i7-11850H 2.50GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.300-preview.24101.10
  [Host]     : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI DEBUG
  DefaultJob : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


```
| Method         | Mean      | Error    | StdDev   | Gen0   | Allocated |
|--------------- |----------:|---------:|---------:|-------:|----------:|
| SingleVariable |  70.49 ns | 1.214 ns | 1.136 ns | 0.0107 |     136 B |
| JustLiterals   | 145.61 ns | 1.443 ns | 1.350 ns | 0.0215 |     272 B |
| Variables      | 144.05 ns | 0.959 ns | 0.897 ns | 0.0215 |     272 B |
| Function       | 116.73 ns | 1.557 ns | 1.381 ns | 0.0191 |     240 B |
| LongString     | 363.35 ns | 2.838 ns | 2.654 ns | 0.4330 |    5432 B |

Results with the flag set (**with optimization**):
```

BenchmarkDotNet v0.13.12, Windows 11 (10.0.22621.3155/22H2/2022Update/SunValley2)
11th Gen Intel Core i7-11850H 2.50GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.300-preview.24101.10
  [Host]     : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI DEBUG
  DefaultJob : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


```
| Method         | Mean        | Error     | StdDev    | Median      | Gen0   | Allocated |
|--------------- |------------:|----------:|----------:|------------:|-------:|----------:|
| SingleVariable |   0.0004 ns | 0.0013 ns | 0.0011 ns |   0.0000 ns |      - |         - |
| JustLiterals   |  12.0562 ns | 0.0719 ns | 0.0673 ns |  12.0500 ns | 0.0032 |      40 B |
| Variables      |  11.9894 ns | 0.0810 ns | 0.0718 ns |  12.0101 ns | 0.0032 |      40 B |
| Function       |  12.2912 ns | 0.1201 ns | 0.1124 ns |  12.3161 ns | 0.0032 |      40 B |
| LongString     | 206.0906 ns | 2.4311 ns | 2.0301 ns | 206.5404 ns | 0.4158 |    5216 B |
## Checklist

- [x] Test cases added
- [x] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated

### CODE REVIEW

File: src/Compiler/Checking/CheckExpressions.fs [hunk 1]
```diff
@@ -7332,6 +7332,33 @@ and TcInterpolatedStringExpr cenv (overallTy: OverallTy) env m tpenv (parts: Syn
...
+                        f xs ys ((mkString g m (sb.ToString()))::acc)
+                    | SynInterpolatedStringPart.FillExpr(_, _)::xs ->
+                        match ys with
+                        | y::ys -> f xs ys (y::acc)
+                        | _ -> error(Error((0, "FOOBAR"), m)) // TODO XXX wrong error
+                    | _ -> acc
+
+                let args = f parts fillExprs [] |> List.rev
+                assert (args.Length = parts.Length)
+                if args.Length = 4 then
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckExpressions.fs:7351]: Was number 4 chosen because that is the max. number where `String.concat` has a dedicated method overload?

Comment (abonie) [reply src/Compiler/Checking/CheckExpressions.fs:7351]: Yeah, see this https://sharplab.io/#v2:DYLgZgzgNALiCWwoBMQGoA+BlAnhGApgLYB0WMATvAHYDmJAwgPbUDGAhjABQBE7PUAAQ8ARgOGtxPZDwCUQA=== vs https://sharplab.io/#v2:DYLgZgzgNALiCWwoBMQGoA+BlAnhGApgLYB0WMATvAHYDmJAwgPbUDGAhjABQBE7PUAAQ8ARgOGtxPZFII8AlEA=

**Comment by abonie** [comment]
Also, if anyone is curious about same benchmarks but **with server GC**:
No optimization:

| Method         | Mean      | Error    | StdDev   | Gen0   | Allocated |
|--------------- |----------:|---------:|---------:|-------:|----------:|
| SingleVariable |  78.08 ns | 1.021 ns | 0.905 ns | 0.0005 |     136 B |
| JustLiterals   | 157.44 ns | 2.675 ns | 2.502 ns | 0.0010 |     272 B |
| Variables      | 151.16 ns | 1.498 ns | 1.401 ns | 0.0010 |     272 B |
| Function       | 125.50 ns | 1.239 ns | 1.099 ns | 0.0010 |     240 B |
| LongString     | 641.34 ns | 2.861 ns | 2.536 ns | 0.0219 |    5432 B |


With optimization:
| Method         | Mean        | Error     | StdDev    | Median      | Gen0   | Allocated |
|--------------- |------------:|----------:|----------:|------------:|-------:|----------:|
| SingleVariable |   0.0064 ns | 0.0103 ns | 0.0096 ns |   0.0000 ns |      - |         - |
| JustLiterals   |  14.8223 ns | 0.1345 ns | 0.1258 ns |  14.8525 ns | 0.0001 |      40 B |
| Variables      |  14.1441 ns | 0.1240 ns | 0.1099 ns |  14.1558 ns | 0.0001 |      40 B |
| Function       |  14.2070 ns | 0.1159 ns | 0.1028 ns |  14.2344 ns | 0.0001 |      40 B |
| LongString     | 508.0675 ns | 6.9382 ns | 6.1506 ns | 507.0075 ns | 0.0210 |    5216 B |

File: src/Compiler/Checking/CheckExpressions.fs [hunk 2]
```diff
@@ -7336,25 +7363,92 @@ and TcInterpolatedStringExpr cenv (overallTy: OverallTy) env m tpenv (parts: Syn
...
+                    (mkStaticCall_String_Concat3 g m args[0] args[1] args[2], tpenv)
+                elif args.Length = 2 then
+                    (mkStaticCall_String_Concat2 g m args[0] args[1], tpenv)
+                elif args.Length = 1 then
+                    args[0], tpenv
                 else
-                    let tyExprs = percentATys |> Array.map (mkCallTypeOf g m) |> Array.toList
-                    mkArray (g.system_Type_ty, tyExprs, m)
+                    error(InternalError("Mismatch in arg count when lowering to Concat", m))
+            else
```
Comment (brianrourkeboll) [inline src/Compiler/Checking/CheckExpressions.fs:7431]: Sometimes keeping things logically separate helps keep things understandable, but if you wanted to do it in one pass, then maybe some of this could be useful:

```fsharp
[<return: Struct>]
let (|HasFormatSpecifier|_|) (s: string) =
    if
        Regex.IsMatch(
            s,
            """
            (^|[^%])                # Start with beginning of string or any char other than '%'
            (%%)*%                  # followed by an odd number of '%' chars
            [+-0 ]{0,3}             # optionally followed by flags
            (\d+)?                  # optionally followed by width
            (\.\d+)?                # optionally followed by .precision
            [bscdiuxXoBeEfFgGMOAat] # and then a char that determines specifier's type
            """,
            RegexOptions.Compiled ||| RegexOptions.IgnorePatternWhitespace)
    then
        ValueSome HasFormatSpecifier
    else
        ValueNone

[<return: Struct>]
let (|WithTrailingStringSpecifierRemoved|_|) (s: string) =
    if s.EndsWith "%s" then
        let i = s.AsSpan(0, s.Length - 2).LastIndexOfAnyExcept '%'
        let diff = s.Length - 2 - i
        if diff &&& 1 <> 0 then ValueSome s[..i] else ValueNone
    else
        ValueNone
```

```fsharp
let rec concatenable acc parts =
    match parts with
    | [] ->
        acc

    | SynInterpolatedStringPart.String (("" | WithTrailingStringSpecifierRemoved ""), _) :: parts ->
        concatenable acc parts

    | SynInterpolatedStringPart.String (s & WithTrailingStringSpecifierRemoved trimmed, m) :: parts ->
        concatenable (SynInterpolatedStringPart.String (s, shiftEnd 0 (trimmed.Length - s.Length) m) :: acc) parts

    | SynInterpolatedStringPart.String (HasFormatSpecifier, _) :: _
    | SynInterpolatedStringPart.FillExpr (_, Some _) :: _
    | SynInterpolatedStringPart.FillExpr (SynExpr.Tuple (isStruct = false; exprs = [_; SynExpr.Const (SynConst.Int32 _align, _)]), _) :: _ ->
        []

    | part :: parts ->
        concatenable (part :: acc) parts

// Lang version check, etc.

match List.rev (concatenable [] parts) with
| [p1; p2; p3; p4] -> mkStaticCall_String_Concat4 g m p1 p2 p3 p4, tpenv
| [p1; p2; p3] -> mkStaticCall_String_Concat3 g m p1 p2 p3, tpenv
| [p1; p2] -> mkStaticCall_String_Concat2 g m p1 p2, tpenv
| [p1] -> p1, tpenv
| _ ->
    // Otherwise...
```

Either way, I've always wanted a good reason to use [`RegexOptions.IgnorePatternWhitespace`](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options#ignore-white-space) and never had one, but this seems like a good opportunity :)

I think you'd need to add a polyfill like this over in [`ReadOnlySpan.fs`](https://github.com/dotnet/fsharp/blob/0f520bce120bb2561f67e467646f4abbdb89d60c/src/Compiler/Utilities/ReadOnlySpan.fs#L8):

```fsharp
[<Extension>]
static member LastIndexOfAnyExcept(span: ReadOnlySpan<char>, value: char) =
    let mutable i = span.Length - 1
    let mutable found = false

    while not found && i >= 0 do
        let c = span[i]

        if c <> value then
            found <- true
        else
            i <- i - 1

    if found then i else -1
```

Comment (abonie) [reply src/Compiler/Checking/CheckExpressions.fs:7431]: I didn't know about `IgnorePatternWhitespace`, that's a way cleaner way to document a regex, thanks!

Regarding the rest, I will have to mull it over - I find it less readable at first glance, partially because it is single sweep, partially I think because I am actually still getting used to active patterns. But it is a neat and concise solution, thanks for the feedback :)

**Review by vzarytovskii — APPROVED** [review]

File: src/Compiler/FSComp.txt [hunk 3]
```diff
@@ -1594,6 +1594,7 @@ featureWarningIndexedPropertiesGetSetSameType,"Indexed properties getter and set
 featureChkTailCallAttrOnNonRec,"Raises warnings if the 'TailCall' attribute is used on non-recursive functions."
 featureUnionIsPropertiesVisible,"Union case test properties"
 featureBooleanReturningAndReturnTypeDirectedPartialActivePattern,"Boolean-returning and return-type-directed partial active patterns"
+featureLowerInterpolatedStringToConcat,"Optimizes interpolated strings in certains cases, by lowering to concatenation"
```
Comment (psfinaki) [inline src/Compiler/FSComp.txt:1597]:
```suggestion
featureLowerInterpolatedStringToConcat,"Optimizes interpolated strings in certain cases, by lowering to concatenation"
```

File: tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs [hunk 4]
```diff
@@ -128,4 +128,34 @@ type Foo () =
         x
         """
         |> compile
-        |> shouldSucceed
\ No newline at end of file
+        |> shouldSucceed
+
+    [<Theory>]
+    // Test different number of interpolated string parts
+    [<InlineData("$\"\"\"abc{\"d\"}e\"\"\"", "abcde")>]
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs:135]: Well if the result is the same for all the theory cases, you don't have to put it in the theory cases :)

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs [hunk 5]
```diff
@@ -34,3 +34,22 @@ IL_0017:  ret"""]
 
 #endif
 
+    [<Fact>]
+    let ``Interpolated string with 3 parts consisting only of strings is lowered to concat`` () =
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs:38]: I think there are more testing opportunities here:
- 1, 2, and 4 parts
- negative cases for more parts
- literals and functions also?

File: src/Compiler/Checking/CheckFormatStrings.fs [hunk 6]
```diff
@@ -317,13 +317,10 @@ let parseFormatStringInternal
...
             |> Seq.length
         if delimLen <= 1 && fmt[i..(i+1)] = "%%" then
-            match context with
-            | Some _ ->
-                specifierLocations.Add(
-                    (Range.mkFileIndexRange m.FileIndex
-                        (Position.mkPos fragLine fragCol)
-                        (Position.mkPos fragLine (fragCol+2))), 0)
-            | None -> ()
+            specifierLocations.Add(
```
Comment (psfinaki) [inline src/Compiler/Checking/CheckFormatStrings.fs:320]: Why are conditions deleted here?

**Review by psfinaki — COMMENTED** [review]
Good and clear PR :)

A question for my education. Ideologically, how is this optimization different from the string optimization that Optimizer does in MakeOptimizedSystemStringConcatCall?

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs (hunk 5 above)
Comment (abonie) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs:38]: Perhaps. Negative case with more parts doesn't really add much value for IL tests - there is a test elsewhere making sure it provides expected results in terms of behavior. Benchmarks kinda confirm that the optimization is applied in those other cases now, but I guess I cannot rule out a weird corner case regression in the future.

File: src/Compiler/Checking/CheckFormatStrings.fs (hunk 6 above)
Comment (abonie) [reply src/Compiler/Checking/CheckFormatStrings.fs:320]: Well, long story short, they shouldn't be removed, so good catch and thanks!

File: tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs (hunk 4 above)
Comment (abonie) [reply tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs:135]: That's true, I expected there's going to be more different examples... and actually perhaps there should at least be a multiline one just in case.

**Comment by psfinaki** [comment]
> I am not really familiar with it, but on a quick glance, it is something else entirely - it (makes and) optimizes a call to System.String.Concat (while this change will make it so that there is a call to Concat in the first place)

Well yeah I just noticed that it also deals with <5 parts so I wonder if your PR doesn't somehow supersede that thing.

File: src/Compiler/Checking/CheckFormatStrings.fs (hunk 6 above)
Comment (psfinaki) [reply src/Compiler/Checking/CheckFormatStrings.fs]: Also, happy I helped, FYI - this is in two places, one is still there :)

**Comment by psfinaki** [comment]
Alright then, thanks for the explanation there :)

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs (hunk 5 above)
Comment (psfinaki) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs:38]: Yeah well primarily I just thought those would be easy to add

File: src/Compiler/Checking/CheckExpressions.fs [hunk 7]
```diff
@@ -140,6 +140,33 @@ exception StandardOperatorRedefinitionWarning of string * range
...
+//----------------------------------------------------------------------------------------------
+module private ParseString =
+    open System.Text.RegularExpressions
+
+    // Regex for strings ending with unescaped '%s'
+    // For efficiency sake using lookbehind, starting from end of the string
+    let stringSpecifier = Regex(@"$(?<=(^|[^%])(%%)*%s)", RegexOptions.Compiled)
+    let EndsWithStringSpecifierNoFlags = stringSpecifier.IsMatch
+
+    // Regex pattern for something like: %[flags][width][.precision][type]
```
Comment (psfinaki) [inline src/Compiler/Checking/CheckExpressions.fs:156]: This is a nice RegEx breakdown - still more examples always help in this cases, so if you put a few examples here in the comments (e.g. with and without optional parts), that would be awesome.

File: src/Compiler/Checking/CheckExpressions.fs [hunk 8]
```diff
@@ -7336,25 +7363,92 @@ and TcInterpolatedStringExpr cenv (overallTy: OverallTy) env m tpenv (parts: Syn
...
-
-            let argsExpr = mkArray (g.obj_ty, fillExprsBoxed, m)
-            let percentATysExpr =
-                if percentATys.Length = 0 then
-                    mkNull m (mkArrayType g g.system_Type_ty)
+            // If all fill expressions are strings and there is less then 5 parts of the interpolated string total
+            // then we can use System.String.Concat instead of a sprintf call.
+            // First filter out all string parts that are empty or only have a '%s' string specifier,
+            // also get rid of the '%s' specifier right before the interpolation expression.
+            let nonEmptyParts =
```
Comment (psfinaki) [inline src/Compiler/Checking/CheckExpressions.fs:7370]: Maybe could be turned into a function to not evaluate this when the feature is not enabled for example.

**Review by psfinaki — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs (hunk 5 above)
Comment (abonie) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/StringFormatAndInterpolation.fs:55]: Added some 👍

File: src/Compiler/Checking/CheckExpressions.fs (hunk 8 above)
Comment (abonie) [reply src/Compiler/Checking/CheckExpressions.fs]: I decided to use another suggestion how to refactor and it solves this issue as well

**Review by psfinaki — APPROVED** [review]

File: src/Compiler/Checking/CheckExpressions.fs [hunk 9]
```diff
@@ -140,6 +141,43 @@ exception StandardOperatorRedefinitionWarning of string * range
...
+let (|WithTrailingStringSpecifierRemoved|) (s: string) =
+    if s.EndsWith "%s" then
+        let i = s.AsSpan(0, s.Length - 2).LastIndexOfAnyExcept '%'
+        let diff = s.Length - 2 - i
+        if diff &&& 1 <> 0 then
+            s[..i]
+        else
+            s
+    else
+        s
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckExpressions.fs:179]: Not necessarily for this particular PR, but I wonder if we can/should accept/return spans here, to not copy strings.

Comment (abonie) [reply src/Compiler/Checking/CheckExpressions.fs:179]: I might be confusing something here, but this only creates a copy of a string if we are slicing it (in `s[..i]` on line 175), right?
If this function accepted and returned a span, we would still have to convert back to a string when calling `mkString`, so would we gain anything?
Am I wrong here, or are you suggesting to change the other APIs to accept spans as well?

Comment (vzarytovskii) [reply src/Compiler/Checking/CheckExpressions.fs:179]: I would suggest changing majority of our non-public APIs to work over spans eventually.

Comment (brianrourkeboll) [reply src/Compiler/Checking/CheckExpressions.fs:179]: > There are literally 0 reasons of using strings over of spans where we can.

Spans are great, but the ergonomics aren't always as good right now — we don't yet support pattern-matching on spans, for example, and you can't return a span from a partial active pattern (you can't do `Some span`/`ValueSome span`).

Now, if we ever get around to type-directed `[]` and a spread operator...

```fsharp
match "abc123".AsSpan 1 with
| ['b'; 'c'; '1'; ..] -> …
| _ -> …
```

or even just being able to match spans against string or array literals (by, say, invoking `op_Implicit`)

```fsharp
match "abc123".AsSpan (1, 3) with
| "bc1" -> …
| _ -> …
```

— _then_ it's really over for strings :)

Comment (vzarytovskii) [reply src/Compiler/Checking/CheckExpressions.fs:179]: Matching on strings or sequences in compiler is mostly on constants, and not something we return as substring of something usually. I'm mostly talking about cases in checker when on checks we need to return a slice of something, and no we just allocate string. We don't need to in most of the cases, I think.

Comment (brianrourkeboll) [reply src/Compiler/Checking/CheckExpressions.fs:179]: > We don't need to in most of the cases, I think.

Yes, in general I agree that there are a lot of places where spans could be used instead of strings.

Comment (brianrourkeboll) [reply src/Compiler/Checking/CheckExpressions.fs:179]: (I went ahead and logged https://github.com/fsharp/fslang-suggestions/issues/1351 anyway 🙃)

Comment (abonie) [reply src/Compiler/Checking/CheckExpressions.fs:179]: I think I will opt not to change it to span for now - in addition to pattern matching, it seems there are some problems using spans with regular expressions (in netstandard2.0). I'd rather tackle this when we decide to change our syntax and typed tree expression to use spans (I really really doubt that usage of these two functions will grow to the point where it is a problem)
